### PR TITLE
implementation for https://github.com/doctrine/mongodb-odm/issues/1106

### PR DESCRIPTION
**Pull request by wodka**
Set typeClass once owner information is passed to collection

-> this does not require any change outside of the Persistent Collection
-> also fixed usage of ODM in https://github.com/sonata-project/SonataAdminBundle/pull/2985

### CODE REVIEW

**Comment by malarzm** [comment]
@wodka: Good approach! You need to check if `targetClass` is set as ODM doesn't require that property, I think it's best to throw an exception saying that `getTypeClass()` is not available for mixed document references. Given that I'd also add a test :)

**Comment by wodka** [comment]
well, looks also better on travis now :)

File: lib/Doctrine/ODM/MongoDB/PersistentCollection.php [hunk 1]
```diff
@@ -255,6 +260,10 @@ public function setOwner($document, array $mapping)
     {
         $this->owner = $document;
         $this->mapping = $mapping;
+
+        if (!empty($this->mapping['targetDocument'])) {
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/PersistentCollection.php]: can you change this to `if ( ! empty($this->mapping['targetDocument'])) {` (for the sake of whitespaces)?

File: lib/Doctrine/ODM/MongoDB/PersistentCollection.php [hunk 2]
```diff
@@ -330,13 +339,24 @@ public function getOwner()
...
     }
 
+    /**
+     * @return ClassMetadata
+     * @throws MongoDBException
+     */
     public function getTypeClass()
     {
+        if (empty($this->typeClass)) {
+            throw new MongoDBException('targetDocument option is required for typeClass');
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/PersistentCollection.php]: Can we rephrase this to "Specifying targetDocument is required for the ClassMetadata to be obtained."?

**Comment by malarzm** [comment]
@wodka I was about to write about my notes but I see you've already adjusted that :) Could you also add a test case if the exception is thrown when `targetDocument` is not set?

**Comment by wodka** [comment]
Done :)

**Comment by malarzm** [comment]
Merged manually in https://github.com/doctrine/mongodb-odm/commit/5b30e0f7a677c8a7494d82176bb8a747f4e82fb9 - thanks @wodka!

**Comment by jmikola** [comment]
@wodka, @malarzm: Was this method present in an interface, or is it just something in ORM that we're implementing for compatibility?

Since this is a new feature, it may be worth adding a changelog entry for it.

**Comment by jmikola** [comment]
Scratch that. I thought this was adding a new method for some reason. It is definitely just a bug fix :)
